python/ray/dashboard/modules/node/node_head.py

```diff
@@ -170,104 +157,122 @@ def get_internal_states(self):
             "head_node_registration_time_s": self._head_node_registration_time_s,
             "registered_nodes": len(DataSource.nodes),
             "registered_agents": len(DataSource.agents),
-            "node_update_count": self._node_update_cnt,
             "module_lifetime_s": time.time() - self._module_start_time,
         }
 
-    async def _get_nodes(self):
-        """Read the client table.
+    async def _subscribe_nodes(self) -> AsyncGenerator[dict, None]:
+        """
+        Yields the initial state of all nodes, then yields the updated state of nodes.
 
-        Returns:
-            A dict of information about the nodes in the cluster.
+        It makes GetAllNodeInfo call only once after the subscription is done, to get
+        the initial state of the nodes.
         """
-        try:
-            nodes = await self.get_all_node_info(timeout=GCS_RPC_TIMEOUT_SECONDS)
-            return {
-                node_id.hex(): gcs_node_info_to_dict(node_info)
-                for node_id, node_info in nodes.items()
-            }
-        except Exception:
-            logger.exception("Failed to GetAllNodeInfo.")
-            raise
+        gcs_addr = self._gcs_address
+        subscriber = GcsAioNodeInfoSubscriber(address=gcs_addr)
+        await subscriber.subscribe()
+
+        # Get all node info from GCS. For TOCTOU, it happens after the subscription.
```

Comment:
What's TOCTOU?

Reply:
if you do:

1. get all
2. subscribe
3. for each update, ...

between 1 and 2 you can miss some nodes. so we do

1. subscribe
2. get all
3. for each update, ...

https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use#:~:text=In%20software%20development%2C%20time%2Dof,the%20results%20of%20that%20check.

Reply:
I think actor_head is not doing that, it's doing `get all` and then `subscribe` so it has the TOCTOU problem?